### PR TITLE
Extend Schedule Pallet for Root Scheduling/Canceling

### PR DESCRIPTION
**Pull request by shawntabrizi**
The scheduler pallet can be used to dispatch calls that would otherwise run into issues with weight limits, for example a runtime upgrade with `set_code`.

This PR introduces dispatch functions that allow the Root origin to add and remove from the scheduler.

This means that if you want to properly do a runtime upgrade using Sudo, you would:

* Sudo Key -> Sudo Call -> Scheduler -> System Set Code

### CODE REVIEW

File: frame/scheduler/src/lib.rs [hunk 1]
```diff
@@ -171,51 +228,48 @@ decl_module! {
...
+	fn do_schedule_named(
+		id: Vec<u8>,
 		when: T::BlockNumber,
 		maybe_periodic: Option<schedule::Period<T::BlockNumber>>,
 		priority: schedule::Priority,
 		call: <T as Trait>::Call,
-	) -> Result<Self::Address, ()> {
+	) -> Result<TaskAddress<T::BlockNumber>, DispatchError> {
 		// determine id and ensure it is unique
 		let id = id.encode();
```
Comment (thiolliere) [inline frame/scheduler/src/lib.rs:270]: You shouldn't encode again

File: frame/support/src/traits.rs [hunk 2]
```diff
@@ -1226,7 +1226,7 @@ pub mod schedule {
 		///
 		/// - `id`: The identity of the task. This must be unique and will return an error if not.
 		fn schedule_named(
-			id: impl Encode,
+			id: Vec<u8>,
```
Comment (gavofyork) [inline frame/support/src/traits.rs:1229]: why?

Comment (gavofyork) [reply frame/support/src/traits.rs:1229]: in order to be able to mak it dispatchable...

**Comment by gavofyork** [comment]
ci.

File: frame/scheduler/src/lib.rs [hunk 3]
```diff
@@ -161,24 +188,48 @@ decl_module! {
...
+		/// Execute the scheduled calls
+		///
+		/// # <weight>
+		/// - S = Number of already scheduled calls
+		/// - Base Weight: 9.243 + 23.45 * S µs
+		/// - DB Weight:
+		///     - Read: Agenda, Lookup
+		///     - Write: Agenda, Lookup
+		/// - In this case, the base weight is dominated by read/write operations.
+		///   To simplify, we will only account for additional reads/writes.
```
Comment (thiolliere) [inline frame/scheduler/src/lib.rs:215]: what does that mean is `23.45 * S` negligible ?

File: frame/scheduler/src/lib.rs [hunk 4]
```diff
@@ -212,7 +263,7 @@ decl_module! {
 							maybe_id,
 							r.map(|_| ()).map_err(|e| e.error)
 						));
-						result = cumulative_weight;
+						total_weight = cumulative_weight;
```
Comment (thiolliere) [inline frame/scheduler/src/lib.rs:266]: I'm not sure this is correct:
if `cumulative_weight > limit` and there is no HARD_DEADLINE scheduled then only the first scheduled is executed. so I think only the weight to this call should be added.

File: frame/scheduler/src/lib.rs [hunk 5]
```diff
@@ -105,31 +108,135 @@ decl_storage! {
...
+			let base_weight: Weight = T::DbWeight::get().reads_writes(1, 1); // Agenda
+			let mut total_weight: Weight = 0;
 			queued.into_iter()
 				.enumerate()
-				.scan(0, |cumulative_weight, (order, (index, s))| {
-					*cumulative_weight += s.call.get_dispatch_info().weight;
+				.scan(base_weight, |cumulative_weight, (order, (index, s))| {
+					*cumulative_weight = cumulative_weight
+						.saturating_add(s.call.get_dispatch_info().weight)
+						.saturating_add(T::DbWeight::get().writes(1)); // Assume worst case: Lookup
```
Comment (thiolliere) [inline frame/scheduler/src/lib.rs:231]: I think worst case also includes for the a prioding scheduled the Agenda::decode_len and Agenda::append

File: frame/scheduler/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/scheduler/src/lib.rs]: added it back

File: frame/scheduler/src/lib.rs [hunk 6]
```diff
@@ -105,31 +108,145 @@ decl_storage! {
...
+		/// # </weight>
 		fn on_initialize(now: T::BlockNumber) -> Weight {
 			let limit = T::MaximumWeight::get();
 			let mut queued = Agenda::<T>::take(now).into_iter()
 				.enumerate()
 				.filter_map(|(index, s)| s.map(|inner| (index as u32, inner)))
 				.collect::<Vec<_>>();
 			queued.sort_by_key(|(_, s)| s.priority);
-			let mut result = 0;
+			let base_weight: Weight = T::DbWeight::get().reads_writes(1, 1) // Agenda
```
Comment (thiolliere) [inline frame/scheduler/src/lib.rs:224]: In worst case we do `Agenda::take(now)` and `Agenda::append(now + one)`
```suggestion
			let base_weight: Weight = T::DbWeight::get().reads_writes(1, 2) // Agenda
```

**Comment by shawntabrizi** [comment]
bot merge

**Comment by gnunicorn** [comment]
bot merge

**Pull request merged**